plugin/src/main/java/com/playmonumenta/scriptedquests/zones/zone/Zone.java

```diff
@@ -0,0 +1,273 @@
+			throw new Exception("propertyGroups may not be null (but may be empty).");
+		}
+
+		Double[] corners = new Double[6];
+		String name;
+		LinkedHashSet<String> properties = new LinkedHashSet<String>();
+
+		// Load the zone name
+		if (object.get("name") == null ||
+		    object.get("name").getAsString() == null ||
+		    object.get("name").getAsString().isEmpty()) {
+			throw new Exception("Failed to parse 'name'");
+		}
+		name = object.get("name").getAsString();
+
+		// Load the zone location
+		if (object.get("location") == null ||
+		    object.get("location").getAsJsonObject() == null) {
+			throw new Exception("Failed to parse 'location'");
+		}
+		JsonObject locationJson = object.get("location").getAsJsonObject();
+		Set<Entry<String, JsonElement>> entries = locationJson.entrySet();
+		for (Entry<String, JsonElement> ent : entries) {
+			String key = ent.getKey();
+			JsonElement value = ent.getValue();
+			switch (key) {
+			case "x1":
+				corners[0] = value.getAsDouble();
+				break;
+			case "y1":
+				corners[1] = value.getAsDouble();
+				break;
+			case "z1":
+				corners[2] = value.getAsDouble();
+				break;
+			case "x2":
+				corners[3] = value.getAsDouble();
+				break;
+			case "y2":
+				corners[4] = value.getAsDouble();
+				break;
+			case "z2":
+				corners[5] = value.getAsDouble();
+				break;
+			default:
+				throw new Exception("Unknown location key: '" + key + "'");
+			}
+		}
+		for (Double cornerAxis : corners) {
+			if (cornerAxis == null) {
+				throw new Exception("Location prereq must have x1 x2 y1 y2 z1 and z2");
+			}
+		}
+		Vector pos1 = new Vector(corners[0], corners[1], corners[2]);
+		Vector pos2 = new Vector(corners[3], corners[4], corners[5]);
+
+		// Load the zone properties
+		if (object.get("properties") == null ||
+		    object.get("properties").getAsJsonArray() == null) {
+			throw new Exception("Failed to parse 'properties'");
+		}
+		Iterator<JsonElement> iter = object.get("properties").getAsJsonArray().iterator();
+		while (iter.hasNext()) {
+			JsonElement element = iter.next();
+			String propertyName = element.getAsString();
+			applyProperty(propertyGroups, properties, propertyName);
+		}
+
+		return new Zone(layer, pos1, pos2, name, properties);
+	}
+
+	/*
+	 * pos1 and pos2 are used similar to /fill:
+	 * - Both are inclusive coordinates.
+	 * - The minimum/maximum are determined for you.
+	 */
+	public Zone(ZoneLayer layer, Vector pos1, Vector pos2, String name, LinkedHashSet<String> properties) {
+		super(pos1, pos2);
+		mLayer = layer;
+		mName = name;
+		mProperties.addAll(properties);
+	}
+
+	/*
+	 * Reset the fragments of this Zone so they can be recalculated without reloading this zone.
+	 * Used to handle ZoneLayers from other plugins. This should only be called by its ZoneLayer.
+	 */
+	public void reloadFragments() {
+		mFragments.clear();
+
+		ZoneFragment initialFragment = new ZoneFragment(this);
+		mFragments.add(initialFragment);
+	}
+
+	public void invalidate() {
```

Comment:
Make protected. 

Also - I think probably the invalidate of BaseZoneTree should be renamed. They don't seem to do the same thing.

Reply:
The process is:
- Tell zones to discard their fragments (invalidate zone), which plugin-provided zones use to reset themselves to their original state.
- For zones not provided by other plugins, discard all of them from the zone manager. They are still tracked by their fragments.
- Generate a new tree structure, including new fragments for plugin-provided zones. Those zones are valid again, but do not reference their old fragments. This can take a few ticks, but the old tree is still valid.
- Mark the new tree as the active one.
- Now the fragments in the old tree are marked as invalid, forcing players inside them to look for the now correct fragments. It is at this point that the references to the fragments unload, which in turn were still referencing the original zones, which can also unload.

Reply:
Ok, I think I understand. Invalidating (removing players) really is still about unloading, so the name does make sense.